snapshots/native/native.go

```diff
@@ -61,57 +61,54 @@ func NewSnapshotter(root string) (snapshots.Snapshotter, error) {
 //
 // Should be used for parent resolution, existence checks and to discern
 // the kind of snapshot.
-func (o *snapshotter) Stat(ctx context.Context, key string) (snapshots.Info, error) {
-	ctx, t, err := o.ms.TransactionContext(ctx, false)
-	if err != nil {
-		return snapshots.Info{}, err
-	}
-	defer t.Rollback()
-	_, info, _, err := storage.GetInfo(ctx, key)
+func (o *snapshotter) Stat(ctx context.Context, key string) (info snapshots.Info, err error) {
+	err = o.ms.WithTransaction(ctx, false, func(ctx context.Context) error {
+		_, info, _, err = storage.GetInfo(ctx, key)
+		return err
+	})
 	if err != nil {
 		return snapshots.Info{}, err
 	}
 
 	return info, nil
 }
 
-func (o *snapshotter) Update(ctx context.Context, info snapshots.Info, fieldpaths ...string) (snapshots.Info, error) {
-	ctx, t, err := o.ms.TransactionContext(ctx, true)
-	if err != nil {
-		return snapshots.Info{}, err
-	}
-
-	info, err = storage.UpdateInfo(ctx, info, fieldpaths...)
+func (o *snapshotter) Update(ctx context.Context, info snapshots.Info, fieldpaths ...string) (_ snapshots.Info, err error) {
+	err = o.ms.WithTransaction(ctx, true, func(ctx context.Context) error {
+		info, err = storage.UpdateInfo(ctx, info, fieldpaths...)
+		return err
+	})
 	if err != nil {
-		t.Rollback()
-		return snapshots.Info{}, err
-	}
-
-	if err := t.Commit(); err != nil {
 		return snapshots.Info{}, err
 	}
 
 	return info, nil
 }
 
-func (o *snapshotter) Usage(ctx context.Context, key string) (snapshots.Usage, error) {
-	ctx, t, err := o.ms.TransactionContext(ctx, false)
-	if err != nil {
-		return snapshots.Usage{}, err
-	}
-	defer t.Rollback()
-
-	id, info, usage, err := storage.GetInfo(ctx, key)
-	if err != nil {
-		return snapshots.Usage{}, err
-	}
+func (o *snapshotter) Usage(ctx context.Context, key string) (usage snapshots.Usage, err error) {
+	err = o.ms.WithTransaction(ctx, false, func(ctx context.Context) error {
+		var (
+			id   string
+			info snapshots.Info
+		)
 
-	if info.Kind == snapshots.KindActive {
-		du, err := fs.DiskUsage(ctx, o.getSnapshotDir(id))
+		id, info, usage, err = storage.GetInfo(ctx, key)
 		if err != nil {
-			return snapshots.Usage{}, err
+			return err
 		}
-		usage = snapshots.Usage(du)
+
+		if info.Kind == snapshots.KindActive {
```

Comment:
Possible optimization after this PR, move the `fs.DiskUsage` call outside the transaction, it can be really slow

Reply:
It's a small bug, I fixed it directly...